### PR TITLE
EVG-14242: Parallelize downloading test results

### PR DESCRIPTION
**Pull request by julianedwards**
JIRA: https://jira.mongodb.org/browse/EVG-14242

The test results rest routes have long latencies (~5s) due to the fact that they must iterator through a pail bucket and download each file individually. To speed up the process I added some parallelization.

### CODE REVIEW

File: model/test_results_iterator.go [hunk 1]
```diff
@@ -49,47 +57,103 @@ func (i *testResultsIterator) Next(ctx context.Context) bool {
...
-		return false
+	close(idxs)
+	var wg sync.WaitGroup
+	catcher := grip.NewBasicCatcher()
+
+	for j := 0; j < runtime.NumCPU(); j++ {
+		wg.Add(1)
+		go func() {
+			defer func() {
+				var err error
```
Comment (Kimchelly) [inline model/test_results_iterator.go:114]: Is this supposed to be declared here? This seems equivalent to just passing nil for the error to `recovery.HandlePanicWithError`.

I noticed this is also done in the log iterator.

File: model/test_results_iterator.go [hunk 2]
```diff
@@ -49,47 +57,103 @@ func (i *testResultsIterator) Next(ctx context.Context) bool {
...
+				catcher.Add(recovery.HandlePanicWithError(recover(), err))
```
Comment (Kimchelly) [inline model/test_results_iterator.go:115]: Can we add a message as the 3rd argument to `recovery.HandlePanicWithError` so we have context for the source of the error (e.g. "iterating test results in worker")?

File: model/test_results_iterator.go [hunk 3]
```diff
@@ -49,47 +57,103 @@ func (i *testResultsIterator) Next(ctx context.Context) bool {
...
+				wg.Done()
+			}()
+
+			for idx := range idxs {
+				if err := ctx.Err(); err != nil {
+					catcher.Add(err)
+					return
+				} else {
```
Comment (Kimchelly) [inline model/test_results_iterator.go:123]: nit: since the if clause returns, the else is unnecessary.

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: model/test_results_iterator.go (hunk 1 above)
Comment (julianedwards) [reply model/test_results_iterator.go:114]: Yeah it should just be nil, copy and paste error. I will change it in both places.

File: model/test_results_iterator.go (hunk 3 above)
Comment (julianedwards) [reply model/test_results_iterator.go:123]: More copying and pasting, I also changed it in the log iterator code.

File: model/test_results_iterator.go [hunk 4]
```diff
@@ -3,13 +3,18 @@ package model
...
+	"sync"
 
 	"github.com/evergreen-ci/pail"
 	"github.com/mongodb/grip"
 	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
 	"go.mongodb.org/mongo-driver/bson"
 )
 
+const testResultsIteratorBatchSize = 100
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply model/test_results_iterator.go:16]: I did think about this, but there's a few other places where configuring batch size would also be useful but it's hard-coded (e.g. the system metrics batch size, log iterator batch size). I would make it a separate ticket to add config options for batch sizing where it's currently hard-coded constants.

**Review by Kimchelly — APPROVED** [review]

**Pull request merged**